### ISSUE TITLE
断点重训 和 非断点重训 的训练顺序不一致（dataloader顺序不一致），导致实验结果无法复现

### ISSUE THREAD

**Issue by NieSYsc20** [verl-project/verl#3457]
你好，断点重训 和 非断点重训 的训练顺序不一致（dataloader）顺序不一致。
如下：
第一次训练为全程没有断点重训：
  粉色：非断点重训

第二次训练为中间断点重训：
  橙色：断点重训前
  绿色：断点重训后

粉色和橙色的prompt len一致，说明在断点重训之前两次dataloader是一致的
粉色和绿色的prompt len不一致，说明在断点重训之前两次dataloader是不一致的
**这会导致一旦断点重训，训练结果无法复现**

<img width="3154" height="1555" alt="Image" src="https://github.com/user-attachments/assets/0a328380-20bc-4c94-8df0-d1ebd7120ba7" />

<img width="2748" height="1561" alt="Image" src="https://github.com/user-attachments/assets/16883f0a-6191-460b-9d96-027260bb0bed" />

<img width="2666" height="1546" alt="Image" src="https://github.com/user-attachments/assets/8500a27a-dc21-40e9-8571-a5ffe8b069c7" />

**Comment by flishwang** [comment]
https://thinkingmachines.ai/blog/defeating-nondeterminism-in-llm-inference/
Do you use any rollout server with nondeterministic algorithms?
Do you use any sampling methods like DAPO (dynamic sampling) that filter prompts based on the responses?
If both are yes, the prompt length could be different.

In dataloader:
Which dataloader did you use? Did you use dataloader with a Sampler to shuffle the dataset? If yes, did you manually set the seed before shuffling?

**Comment by wutong4012** [comment]
I actually have a similar question.
For example, when I use DAPO for training, there may be multiple generation steps within a single training step.
I'm not sure if the dataloader stores this information, but it appears to be loading data starting from the training step.
For example, if my training step count is 100, the generation step count might be 256. This means that the data for the first 256 steps has already been used, but when I resume training, it starts from step 100, resulting in some data being duplicated.
Is this a problem? If so, how can I resolve it?
@flishwang

**Comment by Kirrito-k423** [comment]
中文：这个问题看起来已经由 #4156 修复，并且相关代码仍在 main 分支。修复包括：在 `create_rl_sampler` 中使用 `torchdata.stateful_dataloader.sampler.RandomSampler` 以便 sampler 状态可恢复；恢复训练时加载 `data.pt` 并按 `global_steps // len(train_dataloader)` 从正确的 epoch 继续训练。PR 讨论中提交者在验证后反馈该修改有效。后续 #5725 还补充处理了刚好在 epoch 边界恢复时 dataloader state 已 exhausted 的情况。

English: This appears to have been fixed by #4156, and the relevant code is still present on `main`. The fix uses `torchdata.stateful_dataloader.sampler.RandomSampler` for resumable sampler state and resumes the training loop from `global_steps // len(train_dataloader)` after restoring the dataloader checkpoint. The PR discussion includes a follow-up verification from the contributor saying the change worked. #5725 also added a guard for resuming exactly at an epoch boundary, where the saved dataloader state is exhausted.

中文：因此建议按已修复关闭。如果当前 main/nightly 仍能复现 dataloader 顺序不一致，请带上完整配置、checkpoint step、`data.shuffle`、`data.seed`、是否自定义 sampler，以及 checkpoint 是 epoch 中间还是边界位置，重新提交问题。

English: Closing this as fixed. If anyone can still reproduce a dataloader order mismatch on a current main/nightly build, please open a new issue with the exact config, checkpoint step, `data.shuffle`, `data.seed`, sampler class if customized, and whether the checkpoint is mid-epoch or exactly at an epoch boundary.
This response was generated by a scheduled Codex GPT5.5 issue processing task.
